File: Leetcode/Python3/51. N-Queens/solution.py
# https://leetcode.com/problems/n-queens/description/


from typing import List, Dict, Set


class Solution:
    def solveNQueens(self, n: int) -> List[List[str]]:
        result = []

        def find_queen(queen: int, diag1: Set, diag2: Set, cols: Dict) -> None:
            if queen == n:  # Found queens potision
                result.append(list(cols.keys()))
                return
            for x in range(n):
                # cols = Queen's column index, diag1 = \ diagonal, diag2 = / diagonal
                if x in cols or queen + x in diag1 or queen - x in diag2:
                    continue
                cols[x] = 1
                diag1.add(queen + x)
                diag2.add(queen - x)
                find_queen(queen + 1, diag1, diag2, cols)
                del cols[x]
                diag1.remove(queen + x)
                diag2.remove(queen - x)

        find_queen(0, set(), set(), dict())

        for board in result:  # Draw board
            for idx in range(len(board)):
                pos = board[idx]
                board[idx] = "." * pos + "Q" + "." * (len(board) - pos - 1)

        return result


a = Solution()
print(a.solveNQueens(1))

File: Leetcode/Typescript/51. N-Queens/solution.ts
// https://leetcode.com/problems/n-queens/description/

function solveNQueens(n: number): string[][] {
  const result: string[][] = [];
  placeQueen(
    n,
    [],
    new Set<number>(),
    new Set<number>(),
    new Set<number>(),
    result
  );
  return result;
}

function placeQueen(
  n: number,
  board: number[],
  cols: Set<number>,
  diag1: Set<number>,
  diag2: Set<number>,
  result: string[][]
) {
  const row = board.length;
  // Placed all queen
  if (row === n) {
    result.push(boardToString(board));
    return;
  }

  for (let col = 0; col < n; col++) {
    // Checks if a queen can be placed in the given column without being attacked by other queens
    if (canPlaceQueen(diag1, diag2, cols, col, row)) {
      board.push(col);
      cols.add(col);
      diag1.add(row + col);
      diag2.add(row - col);
      placeQueen(n, board, cols, diag1, diag2, result);
      board.pop();
      cols.delete(col);
      diag1.delete(row + col);
      diag2.delete(row - col);
    }
  }
}

function boardToString(board: number[]): string[] {
  const result: string[] = [];

  for (const col of board) {
    const rowBoard = Array.from({ length: board.length }, (_) => ".");

    rowBoard[col] = "Q";
    result.push(rowBoard.join(""));
  }

  return result;
}

function canPlaceQueen(
  diag1: Set<number>,
  diag2: Set<number>,
  cols: Set<number>,
  col: number,
  row: number
): boolean {
  if (cols.has(col) || diag1.has(row + col) || diag2.has(row - col)) {
    return false;
  }
  return true;
}
